### PR TITLE
Shank 408 open api

### PR DESCRIPTION
**Pull request by jhulbert-va**
https://vasdvp.atlassian.net/browse/SHANK-408

### CODE REVIEW

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+          description = "Production",
+          variables = @ServerVariable(name = "version", defaultValue = "v0"))
+    })
+@SecurityScheme(
+    paramName = "apikey",
+    type = SecuritySchemeType.APIKEY,
+    name = "apikey",
+    in = SecuritySchemeIn.HEADER)
+@Path("/")
+public interface FacilitiesService
```
Comment (jblefkove-va) [inline facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java]: Is the facilities controller in the main app going to implement this interface?

File: facilities/pom.xml [hunk 2]
```diff
@@ -46,4 +47,60 @@
...
+          <execution>
+            <phase>compile</phase>
+            <configuration>
+              <target>
+                <!--
+              Sadness here.
+              Swagger is ignore @JsonProperty annotations on enums. This big bit of sadness changes generated swagger _ to - in enums.
+            -->
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.yaml" flags="g" match="([a-z])_([a-z])" replace="\1-\2"/>
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.yaml" flags="g" match="- _([a-z])" replace="- \1"/>
```
Comment (jblefkove-va) [inline facilities/pom.xml]: These replacements are for dealing with hyphens vs. underscores in the FHIR types. I don't think we need them there?

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java (hunk 1 above)
Comment (jhulbert-va) [reply facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java]: No

File: facilities/pom.xml [hunk 3]
```diff
@@ -46,4 +47,60 @@
...
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.json" flags="g" match="([a-z])_([a-z])" replace="\1-\2"/>
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.json" flags="g" match="(,|\[) &quot;_([a-z])" replace="\1 &quot;\2"/>
+                <!--
+              This group is because the swagger auto-generated for dates, times, and datetimes examples doesn't play nice with how swagger is actually consumed.
+              The first regex captures datetimes and times. The second one covers left over dates.
+            -->
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.yaml" flags="g" match="\b(([0-9]{4}(-(0[1-9]|1[0-2])(-(0[0-9]|[1-2][0-9]|3[0-1]))?)T)?(([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\.[0-9]+)?)(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))?)\b" replace="'\1'"/>
+                <replaceregexp byline="true" file="${project.build.outputDirectory}/openapi.yaml" flags="g" match="( ([0-9]{4}-(0[1-9]|1[0-2])(-(0[0-9]|[1-2][0-9]|3[0-1]))?))" replace=" '\2'"/>
```
Comment (jblefkove-va) [inline facilities/pom.xml]: IIRC this had something to do with dates not being properly formatted inside our Swagger examples for data-query. Can we remove these as well and verify if that hurts anything?

File: facilities/pom.xml [hunk 4]
```diff
@@ -46,4 +47,60 @@
...
   </dependencies>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>io.swagger.core.v3</groupId>
+        <artifactId>swagger-maven-plugin</artifactId>
+        <version>${swagger-maven-plugin.version}</version>
+        <executions>
+          <execution>
+            <id>data-query-swag</id>
```
Comment (jblefkove-va) [inline facilities/pom.xml]: copypasta

Comment (jhulbert-va) [reply facilities/pom.xml]: :disappointed:

File: facilities/pom.xml (hunk 3 above)
Comment (jhulbert-va) [reply facilities/pom.xml]: Date look fine after removal.

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    tags = @Tag(name = "facilities", description = "VA Facilities API"),
+    servers = {
+      /*  When we switch over to sandbox, we'll need this.
+          @Server(
+              url = "https://sandbox-api.va.gov/services/va_facilities/{version}",
+              description = "Sandbox",
+              variables = @ServerVariable(name = "version", defaultValue = "v0")),
+      */
+      @Server(
+          url = "https://dev-api.va.gov/services/va_facilities/{version}",
```
Comment (jblefkove-va) [inline facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java]: The Swagger in the current API has already switched over from `dev-api` to `sandbox`

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                    + "- Production: `https://api.va.gov/services/va_facilities/v0`"
+                    + "\n\n### Authorization\n\n"
+                    + "API requests are authorized through a symmetric API token, "
+                    + "provided in an HTTP header with name `apikey`."
+                    + "\n\n### Response Formats\n\n"
+                    + "Clients may request several response formats by setting the `Accept` header."
+                    + "\n- application/json - The default JSON response format "
+                    + "complies with JSON API. This media type is *not* available for "
+                    + "bulk requests using the `/facilities/all` endpoint. "
+                    + "It will return `406 Not Acceptable`. .\n"
```
Comment (jblefkove-va) [inline facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java]: Stray period. (I know it's like this in the original)

**Comment by jblefkove-va** [comment]
image

This is wonky. (Current API on the left)

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java (hunk 6 above)
Comment (jhulbert-va) [reply facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/FacilitiesService.java]: > image
> 
> This is wonky. (Current API on the left)



> image
> 
> This is wonky. (Current API on the left)

I capitalized a tag somewhere is my guess.

**Comment by jblefkove-va** [comment]
image
The current API has extra line breaks in some of the longer paragraphs that we probably want as well.

**Comment by jblefkove-va** [comment]
image
Sample 200 response for search doesn't look right

**Comment by jblefkove-va** [comment]
image
The sample 200 response for search doesn't look right.

**Comment by jblefkove-va** [comment]
image

**Comment by jblefkove-va** [comment]
image

**Comment by jhulbert-va** [comment]
https://github.com/department-of-veterans-affairs/lighthouse-facilities/pull/26#issuecomment-603987858

We changed this in favor of a "GenericError" template, not two error models that had the same structure.

**Comment by jblefkove-va** [comment]
@jhulbert-va After some more thought, I realized a way to support upper- and lowercase hours with just one set of fields. I've committed that here... this simplifies  the custom serializer and the avoids having to hide them in the Swagger.

File: facilities/pom.xml [hunk 7]
```diff
@@ -46,4 +47,31 @@
...
+            <goals>
+              <goal>resolve</goal>
+            </goals>
+            <configuration>
+              <resourceClasses>
+                <resourceClass>gov.va.api.lighthouse.facilities.api.FacilitiesService</resourceClass>
+              </resourceClasses>
+              <outputFileName>openapi</outputFileName>
+              <outputPath>${project.build.outputDirectory}</outputPath>
+              <outputFormat>JSONANDYAML</outputFormat>
```
Comment (jblefkove-va) [inline facilities/pom.xml]: Was the outcome of the `LocalDate` problem yesterday that we only want to generate `openapi.json`?

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  public Object openapiJson() throws IOException {
+    return FacilitiesHomeController.MAPPER.readValue(openapiContent(), Object.class);
+  }
+
+  /** OpenAPI Yaml. */
+  @GetMapping(
+      value = {"/openapi.yaml"},
+      produces = "application/vnd.oai.openapi")
+  @ResponseBody
+  public String openapiYaml() throws IOException {
```
Comment (jblefkove-va) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java]: Do we still need to expose yaml?

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  @Bean
+  public String openapiContent() throws IOException {
+    try (InputStream is = openapi.getInputStream()) {
+      return StreamUtils.copyToString(is, Charset.defaultCharset());
+    }
+  }
+
+  /** OpenAPI Json. */
+  @GetMapping(
+      value = {"/", "/openapi.json"},
```
Comment (jblefkove-va) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java]: Also needs to be available at `docs/v0/api`

File: facilities/pom.xml (hunk 7 above)
Comment (jhulbert-va) [reply facilities/pom.xml]: @bschofield-va I believe that the outcome was favoring json over yaml. Not removing the yaml generation.

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java (hunk 9 above)
Comment (jhulbert-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/FacilitiesHomeController.java]: This will require a change to kong as well.

**Pull request merged**